### ISSUE TITLE
Form's OnValidSubmit fires before validation is complete (for Async validation rules)

### ISSUE THREAD

**Issue by afmuller** [Blazored/FluentValidation#31]
I am finding that if
  1. I have async validation rules
  2. The submit button is immediately clicked on the form 
that the OnValidSubmit handler is fired - even though the form is invalid. Only after it has fired, does the ValidationSummary component show the errors.

My Form:
```
        <EditForm Model="@contactModel" OnValidSubmit="@HandleValidSubmit" OnInvalidSubmit="@HandleInvalidSubmit" Context="EditFormContext">
            <FluentValidationValidator />
            <ValidationSummary />

            <p>
                <label>First Name: </label>
                <InputText @bind-Value="@contactModel.FirstName" />
            </p>

            <p>
                <label>Last Name: </label>
                <InputText @bind-Value="@contactModel.LastName" />
            </p>

            <p>
                <label>Tel Number: </label>
                <InputText @bind-Value="@contactModel.TelephoneNumber" />
            </p>

            <p>
                <label>Cell Number: </label>
                <InputText @bind-Value="@contactModel.CellNumber" />
            </p>

            <p>
                <label>Email Address: </label>
                <InputText @bind-Value="@contactModel.EmailAddress" />
            </p>

            <button type="submit">Save</button>
        </EditForm>

```
My validator:
```
    public class CreateContactCommandValidator : AbstractValidator<CreateContactCommand>
    {
        private readonly IContactsAppContext contactsAppContext;

        public CreateContactCommandValidator(IContactsAppContext contactsAppContext)
        {
            RuleFor(x => x.FirstName).NotEmpty();
            RuleFor(x => x.LastName).NotEmpty();
            RuleFor(x => x.EmailAddress).NotEmpty().EmailAddress().MustAsync(BeUnique).WithMessage("Email already registered");

            this.contactsAppContext = contactsAppContext ?? throw new ArgumentNullException(nameof(contactsAppContext));
        }

        async Task<bool> BeUnique(string email, CancellationToken cancellationToke)
        {
            return await contactsAppContext.Contacts.CountAsync(p => p.EmailAddress == email) == 0 ? true : false;
        }
    }

```

In this example, if I click the submit button on an empty form, the OnValidSubmit event is fired, even though the Email Address is empty (and therefore invalid). Only after the OnValidSubmit event has fired, does the ValidationSummary component show the validation errors. 
This only seems to happen if I have async validation rules. If I remove the 'MustAsync' validation rule from the email address, everything works just fine. 

Please advise if I am doing something wrong.
Thanks
Andrew

**Comment by sfmskywalker** [comment]
I'm seeing the exact same thing when I use async validation. I haven't figured out exactly why, but I suspect that somehow the following call doesn't block:

```csharp
EditContext.Validate();
```

When handling `OnValidSubmit`, the above line is of course called by Blazor itself. To test my suspicion, I handled `OnSubmit` using the following handler:

```csharp
private async void OnSubmit()
{
    if (!EditContext.Validate())
        return;

    // The following code always run if I have an async validator in place, even when validation rules return false.
    await SaveDataAsync();
}
```

To work around this issue, I basically need to invoke the FluentValidation myself:

```csharp
private async void OnSubmit()
{
    using var scope = Services.CreateScope();
    var validator = scope.ServiceProvider.GetRequiredService<MyValidator>();
    var result = await validator.ValidateAsync(MyModel);

    if (!result.IsValid)
       return;

    // This code now never runs when validation failed (even when including async rules)
    await SaveDataAsync();
}
```

Although this works for now, it's unfortunate that this doesn't tie into the EditContext stuff directly, which is what FluentValidationValidato provides nicely. In fact, I have to replicate some of its code such as using the `ValidationMessageStore` and populating it with validation errors to get my validation messages on screen.

Long story short, I suspect there's a limitation with Blazor's invocation of EditContext.Validate - probably they need to make it async?

**Comment by chrissainty** [comment]
Hi @afmuller @sfmskywalker 

The problem with async validation is that Blazors validation system is event based and therefore cannot know if async validation is happening and when it's completed. You can see this here: https://github.com/Blazored/FluentValidation/blob/024dbbd990fb25173402986a3072dbc4ded41196/src/Blazored.FluentValidation/EditContextFluentValidationExtensions.cs#L24-L25

This means there is a potential for a form to be submitted before validation completes, as you're seeing. There are two options here, duplicate your validation rules on client and server (I'm assuming you're doing this anyway as you can't rely on client-side validation anyway). This has the potential to return a 400 error from the server with failed validation that end which you can then handle. The other option is to use `OnSubmit` and then validation the form yourself manually before submitting.

I appreciate neither of these are ideal scenarios but this will require a framework change in Blazor to resolve. I'm going to close this issue as there is nothing we can do about it in this library.

**Comment by sfmskywalker** [comment]
Thanks @chrissainty for the explanation, that all makes sense.
I'll go with the other option (using `OnSubmit` and validating the form manually before submitting).
